apps/ensrainbow/download-database-artifact.sh

```diff
@@ -0,0 +1,153 @@
+        return 0 # Success
+    else
+        return 1 # Failure
+    fi
+}
+
+# 1. Download checksum file first
+download_with_progress "$BASE_URL/$SERVER_CHECKSUM_PATH" "$TARGET_CHECKSUM_FILE_PATH" "Checksum file ($DATA_FILE_BASENAME.sha256sum)"
+
+# 2. Check if data file already exists and is valid
+if [ -f "$TARGET_DATA_FILE_PATH" ]; then
+    echo "Data file ($TARGET_DATA_FILE_PATH) already exists."
+    if verify_checksum; then
+        echo "✓ Checksum VERIFIED for existing data file ($DATA_FILE_BASENAME)."
+
+        # Optionally, ensure license file is also present
+        if [ ! -f "$TARGET_LICENSE_FILE_PATH" ]; then
+            echo "License file ($TARGET_LICENSE_FILE_PATH) is missing. Downloading it..."
+            download_with_progress "$BASE_URL/$SERVER_LICENSE_PATH" "$TARGET_LICENSE_FILE_PATH" "License file ($SERVER_LICENSE_PATH)"
+        else
+            echo "License file ($TARGET_LICENSE_FILE_PATH) already exists."
+        fi
+
+        echo "All required files are present and valid."
+        echo "  Data:     $TARGET_DATA_FILE_PATH"
+        echo "  Checksum: $TARGET_CHECKSUM_FILE_PATH"
+        echo "  License:  $TARGET_LICENSE_FILE_PATH"
+        exit 0
+    else
+        echo "⚠ Checksum FAILED for existing data file ($DATA_FILE_BASENAME)."
+        echo "Will proceed to download a fresh copy of the data file."
+        rm -f "$TARGET_DATA_FILE_PATH" # Remove potentially corrupted existing file
+    fi
+else
+    echo "Data file ($TARGET_DATA_FILE_PATH) not found. Proceeding with download."
+    echo ""
+fi
+
+# 3. Download License File (if not already downloaded and exited above)
+# This check ensures we don't re-download if it was fetched during the "existing valid data" check.
+if [ ! -f "$TARGET_LICENSE_FILE_PATH" ]; then
+    download_with_progress "$BASE_URL/$SERVER_LICENSE_PATH" "$TARGET_LICENSE_FILE_PATH" "License file ($SERVER_LICENSE_PATH)"
+else
+    # This case is mostly for when data file was missing, but license might have been there from a previous partial run.
+    echo "License file ($TARGET_LICENSE_FILE_PATH) already present."
+    echo ""
+fi
+
+# 4. Download Data File (if not already present and valid)
+download_with_progress "$BASE_URL/$SERVER_DATA_PATH" "$TARGET_DATA_FILE_PATH" "Database artifact ($DATA_FILE_BASENAME)"
+
+# 5. Verify downloaded data file
+echo "Verifying checksum of newly downloaded data file ($DATA_FILE_BASENAME)..."
+if verify_checksum; then
+    echo "✓ Download successful and checksum VERIFIED for $DATA_FILE_BASENAME!"
+else
+    echo "❌ CRITICAL ERROR: Checksum FAILED after download for $DATA_FILE_BASENAME using $(basename "$TARGET_CHECKSUM_FILE_PATH")."
+    echo "The downloaded data file may be corrupted or incomplete."
+    # Consider cleaning up $TARGET_DATA_FILE_PATH here as well
+    exit 1
+fi
+
+echo ""
+echo "---------------------------------------------------"
+echo "ENS Database Artifact download and verification complete."
```

Comment:
```suggestion
echo "ENS Database Artifact download and checksum verification complete."
```

Suggesting this as we only verified the checksum. In theory it might still not be a valid database for ENSRainbow.